File: admin_web/src/lib/marketplace-people-contract.test.ts
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { test } from "node:test";

import { translateValue } from "./french-dom.ts";

const source = readFileSync(
  new URL("../components/admin-console.tsx", import.meta.url),
  "utf8",
);
const businessPeopleSource = readFileSync(
  new URL(
    "../components/business/profile-support-people.tsx",
    import.meta.url,
  ),
  "utf8",
);

test("web people directory uses the sanitized paginated callable", () => {
  assert.match(source, /"listMarketplacePeople"/);
  assert.match(source, /nextPageToken|pageToken/);
  assert.match(source, /Load more people/);
  assert.doesNotMatch(source, /loadEveryAuthPage/);
  assert.match(source, /getMarketplacePerson/);
  assert.doesNotMatch(
    source,
    /collection\(db,\s*"users"\)[\s\S]{0,160}(?:onSnapshot|getDocs)/,
  );
  assert.doesNotMatch(
    source,
    /"users",\s*tabNeeds\([^)]*"people"/,
    "People must not enable the raw users collection listener",
  );
});

test("unverified administrators get a bilingual recovery gate before people data or invitations", () => {
  assert.match(source, /sendEmailVerification/);
  assert.match(source, /user\.reload\(\)/);
  assert.match(source, /getIdToken\(true\)/);
  assert.match(
    source,
    /useAdminAuthUsers\(\s*tabNeeds\("people"\) && adminEmailVerified,\s*\)/,
  );
  assert.match(source, /if \(!adminEmailVerified\)/);
  assert.match(source, /verificationBusy === "sending"/);
  assert.match(source, /verificationBusy === "checking"/);

  for (const label of [
    "Security check required",
    "Verify your admin email to manage people",
    "Send verification email",
    "I verified — check again",
    "People management is locked",
    "Roles & security",
  ]) {
    assert.notEqual(
      translateValue(label, "fr"),
      label,
      `missing French translation for "${label}"`,
    );
  }
});

test("web exposes unified categories and guarded lifecycle actions", () => {
  for (const category of [
    "Platform administrators",
    "Business owners",
    "Business staff",
    "Customers",
    "Pending invitations",
    "Missing profiles",
  ]) {
    assert.match(source, new RegExp(category));
  }

  for (const callable of [
    "setMarketplaceUserStatus",
    "revokeUserSessions",
    "sendUserRecoveryEmail",
    "invitePlatformAdmin",
    "inviteBusinessMember",
    "resendAccessInvitation",
    "cancelAccessInvitation",
    "transferBusinessOwnership",
    "reviewAccountDeletion",
    "finalizeAccountDeletion",
  ]) {
    assert.match(source, new RegExp(`"${callable}"`));
  }
  assert.match(source, /data-loading|disabled=\{[^}]*loading|runAction/);
});

test("web keeps sensitive actions unavailable for self and insufficient access", () => {
  assert.match(source, /currentUserId/);
  assert.match(source, /isCurrentUser/);
  assert.match(source, /canManage/);
  assert.match(source, /businessPermissions/);
  assert.match(source, /businessId/);
  assert.match(source, /last super admin|last business owner/i);
});

test("new user-management copy is bilingual", () => {
  for (const label of [
    "Platform administrators",
    "Business owners",
    "Business staff",
    "Customers",
    "Pending invitations",
    "Missing profiles",
    "Suspend account",
    "Restore account",
    "Revoke sessions",
    "Send password reset",
    "Send verification email",
    "Transfer ownership",
    "Resend invitation",
    "Cancel invitation",
    "Review deletion request",
    "Finalize account deletion",
  ]) {
    assert.notEqual(
      translateValue(label, "fr"),
      label,
      `missing French translation for "${label}"`,
    );
  }
});

test("every web personnel entry point uses invitations, never shared passwords", () => {
  const personnelSources = `${source}\n${businessPeopleSource}`;
  assert.match(businessPeopleSource, /"inviteBusinessMember"/);
  assert.doesNotMatch(personnelSources, /"createStaffUser"/);
  assert.doesNotMatch(personnelSources, /"createPlatformManager"/);
  assert.doesNotMatch(
    businessPeopleSource,
    /Temporary password|Share the temporary password/,
  );
  assert.match(businessPeopleSource, /businessPermissions/);
  assert.match(source, /people-permission-picker/);
  assert.match(source, /locale:\s*currentInterfaceLocale\(\)/);
  assert.match(source, /await refreshPeople\(\)/);
});

test("the preview administrator reports one consistent super-admin role", () => {
  assert.match(
    source,
    /id:\s*"admin-preview"[\s\S]{0,240}adminRole:\s*"superAdmin"/,
  );
});

File: my_flutter_app/test/marketplace_people_contract_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

void main() {
  final screen = File(
    'lib/screens/user_management_screen.dart',
  ).readAsStringSync();
  final provider = File('lib/providers/auth_provider.dart').readAsStringSync();
  final staffInvite = File(
    'lib/screens/add_staff_screen.dart',
  ).readAsStringSync();
  final english = File('lib/l10n/app_en.arb').readAsStringSync();
  final french = File('lib/l10n/app_fr.arb').readAsStringSync();

  test('mobile uses the same sanitized marketplace directory contract', () {
    expect('$screen\n$provider', contains('listMarketplacePeople'));
    expect('$screen\n$provider', contains('getMarketplacePerson'));
    expect(screen, isNot(contains("collection('users').snapshots()")));
    expect(screen, contains('pageToken'));
  });

  test('mobile exposes the same safe lifecycle callables as web', () {
    for (final callable in <String>[
      'setMarketplaceUserStatus',
      'revokeUserSessions',
      'sendUserRecoveryEmail',
      'invitePlatformAdmin',
      'inviteBusinessMember',
      'resendAccessInvitation',
      'cancelAccessInvitation',
      'transferBusinessOwnership',
      'reviewAccountDeletion',
      'finalizeAccountDeletion',
    ]) {
      expect(
        '$screen\n$provider',
        contains(callable),
        reason: '$callable must be shared with the web management contract',
      );
    }
  });

  test('mobile renders unified categories and permission-aware actions', () {
    for (final token in <String>[
      'platformAdministrators',
      'businessOwners',
      'businessStaff',
      'customers',
      'pendingInvitations',
      'missingProfiles',
      'businessPermissions',
      'businessId',
    ]) {
      expect(screen, contains(token));
    }
    expect(screen, contains('isCurrentUser'));
    expect(screen, contains('canManage'));
  });

  test('all marketplace people copy exists in both ARB catalogs', () {
    for (final key in <String>[
      'platformAdministrators',
      'businessOwners',
      'businessStaff',
      'pendingInvitations',
      'missingProfiles',
      'suspendAccount',
      'restoreAccount',
      'revokeSessions',
      'sendPasswordReset',
      'sendVerificationEmail',
      'transferOwnership',
      'resendInvitation',
      'cancelInvitation',
      'reviewDeletionRequest',
      'finalizeAccountDeletion',
    ]) {
      expect(english, contains('"$key"'));
      expect(french, contains('"$key"'));
    }
  });

  test(
    'mobile personnel entry points use invitations, not shared passwords',
    () {
      final personnelSources = '$screen\n$provider\n$staffInvite';
      expect(personnelSources, isNot(contains("'createStaffUser'")));
      expect(personnelSources, isNot(contains("'createPlatformManager'")));
      expect(staffInvite, contains('inviteBusinessMember'));
      expect(staffInvite, contains('businessPermissions'));
      expect(staffInvite, isNot(contains('Temporary password')));
    },
  );
}
